<a href="https://colab.research.google.com/github/polina-minaeva/python-proj/blob/main/%D0%98%D1%82%D0%BE%D0%B3%D0%BE%D0%B2%D1%8B%D0%B9_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82_Python_%D0%B4%D0%BB%D1%8F_%D0%B0%D0%BD%D0%B0%D0%BB%D0%B8%D0%B7%D0%B0_%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D1%85_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#загружаем библиотеки
import pandas as pd
import matplotlib.pyplot as plt

1) Загрузим файл HR.csv в pandas dataframe

In [ ]:
df_survey = pd.read_csv('HR.csv') #загружаем файл HR.csv в pandas dataframe
df_survey.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [ ]:
df_survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   department             14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


2) Рассчитаем основные статистики для переменных
(среднее,медиана,мода,мин/макс,сред.отклонение)

In [ ]:
#отберем только столбы с количественными данными
df_survey_quant = df_survey[['satisfaction_level', 'last_evaluation', 'number_project', 'average_montly_hours', 'time_spend_company']]
df_survey_quant.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  
0                   3  
1                   6  
2                   4  
3                   5  
4                   3

In [ ]:
#также отберем только столбы с качественными данными
df_survey_qual = df_survey[['Work_accident', 'left', 'promotion_last_5years', 'department', 'salary']]
df_survey_qual.head()

Work_accident  left  promotion_last_5years department  salary
0              0     1                      0      sales     low
1              0     1                      0      sales  medium
2              0     1                      0      sales  medium
3              0     1                      0      sales     low
4              0     1                      0      sales     low

In [ ]:
'''рассчитаем основные статистики для переменных
(среднее,медиана,мода,мин/макс,сред.отклонение).'''
'''для количественных - среднее, медиану, моду, мин/макс, ср.откл.
для качественных - моду'''

df_survey_quant.mean()

satisfaction_level        0.612834
last_evaluation           0.716102
number_project            3.803054
average_montly_hours    201.050337
time_spend_company        3.498233
dtype: float64

In [ ]:
df_survey_quant.median ()

satisfaction_level        0.64
last_evaluation           0.72
number_project            4.00
average_montly_hours    200.00
time_spend_company        3.00
dtype: float64

In [ ]:
df_survey_quant.mode()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                 0.1             0.55             4.0                   135   
1                 NaN              NaN             NaN                   156   

   time_spend_company  
0                 3.0  
1                 NaN

In [ ]:
df_survey_quant.describe() #узнаем мин, макс, ср.откл.

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  
count          14999.000000        14999.000000  
mean             201.050337            3.498233  
std               49.943099            1.460136  
min               96.000000            2.000000  
25%              156.000000            3.000000  
50%              200.000000            3.000000  
75%              245.000000            4.000000  
max              310.000000           10.000000

По данным мы видим, что, в среднем, работники делают 3,8 проекта. Среднее количество рабочих часов - 201. Сотрудники, в среднем, имеют стаж 3,5 года. В среднем, уровень удовлетворенности работой равен 0,6.

In [ ]:
df_survey_qual.mode()

Work_accident  left  promotion_last_5years department salary
0              0     0                      0      sales    low

По данным мы видим, что чаще всего в компании работники получают низкую зарплату. Большинство сотрудников задействованы в департаменте sales. Не уволившихся больше, чем уволившихся. 

3) Рассчитаем и визуализируем корреляционную матрицу для количественных переменных. Определим две самые скоррелированные и две наименее скоррелированные переменные

In [ ]:
df_survey_quant.corr() #создаем корреляционную матрицу

satisfaction_level  last_evaluation  number_project  \
satisfaction_level              1.000000         0.105021       -0.142970   
last_evaluation                 0.105021         1.000000        0.349333   
number_project                 -0.142970         0.349333        1.000000   
average_montly_hours           -0.020048         0.339742        0.417211   
time_spend_company             -0.100866         0.131591        0.196786   

                      average_montly_hours  time_spend_company  
satisfaction_level               -0.020048           -0.100866  
last_evaluation                   0.339742            0.131591  
number_project                    0.417211            0.196786  
average_montly_hours              1.000000            0.127755  
time_spend_company                0.127755            1.000000

In [ ]:
vis_corr = df_survey_quant.corr() #визуализируем матрицу с помощью style от matplotlip
vis_corr.style.background_gradient(cmap='RdYlGn')

На визуализации мы видим, что две наиболее скоррелированные переменные - average_montly_hours и number_project. Их общая ячейка (0.417211) самая светлая. Этому соотвествует предположение, что количество выполненных проектов зависит от рабочих часов в месяц. В свою очередь, две наименее скоррелированные переменные - number_project и satisfaction_level. Их общая ячейка (-0.142970) самая бордовая. На основании этого можно предположить, что большое количество проектов негативно сказывается на уровне удовлетворенности работой.

4) Рассчитаем сколько сотрудников работает в каждом департаменте

In [ ]:
#узнаем сколько сотрудников работает в каждом департаменте
df_survey.groupby(['department']).count()

satisfaction_level  last_evaluation  number_project  \
department                                                         
IT                         1227             1227            1227   
RandD                       787              787             787   
accounting                  767              767             767   
hr                          739              739             739   
management                  630              630             630   
marketing                   858              858             858   
product_mng                 902              902             902   
sales                      4140             4140            4140   
support                    2229             2229            2229   
technical                  2720             2720            2720   

             average_montly_hours  time_spend_company  Work_accident  left  \
department                                                                   
IT                           1227                1227           1227  1227   
RandD                         787                 787            787   787   
accounting                    767                 767            767   767   
hr                            739                 739            739   739   
management                    630                 630            630   630   
marketing                     858                 858            858   858   
product_mng                   902                 902            902   902   
sales                        4140                4140           4140  4140   
support                      2229                2229           2229  2229   
technical                    2720                2720           2720  2720   

             promotion_last_5years  salary  
department                                  
IT                            1227    1227  
RandD                          787     787  
accounting                     767     767  
hr                             739     739  
management                     630     630  
marketing                      858     858  
product_mng                    902     902  
sales                         4140    4140  
support                       2229    2229  
technical                     2720    2720

Мы видим, что большинство сотрудников работают в департаменте sales. Также большое их количество работают в департаментах technical и support. Меньше всего работников задействованы в management, hr и accounting.

5) Сделаем распределение сотрудников по зарплатам

In [ ]:
'''посмотрим распределение сотрудников по зарплатам'''
df_survey_all_sal = df_survey.groupby(['salary']).count()
df_survey_all_sal

satisfaction_level  last_evaluation  number_project  \
salary                                                        
high                  1237             1237            1237   
low                   7316             7316            7316   
medium                6446             6446            6446   

        average_montly_hours  time_spend_company  Work_accident  left  \
salary                                                                  
high                    1237                1237           1237  1237   
low                     7316                7316           7316  7316   
medium                  6446                6446           6446  6446   

        promotion_last_5years  department  
salary                                     
high                     1237        1237  
low                      7316        7316  
medium                   6446        6446

Мы видим, что большинство сотрудников получают низкую зарплату. Высокую получают меньшинство.

7) Проверим гипотезу, что сотрудники с высоким окладом проводят на работе больше времени, чем сотрудники с низким окладом

In [ ]:
df_two_par = df_survey[['average_montly_hours', 'salary']]

In [ ]:
df_low = df_two_par[df_two_par['salary']== 'low']

In [ ]:
list_low = df_low['average_montly_hours'].tolist() #создадим два списка в которые поместим згначения часов работы
                                                    #для зарплат - низких и высоких

In [ ]:
df_high = df_two_par[df_two_par['salary']== 'high']

In [ ]:
list_high = df_high['average_montly_hours'].tolist()

In [ ]:
print(list_low)

[157, 223, 159, 153, 247, 259, 224, 142, 135, 305, 234, 148, 137, 143, 160, 255, 160, 262, 282, 147, 304, 139, 158, 242, 239, 135, 128, 132, 294, 134, 145, 140, 246, 255, 137, 126, 306, 152, 269, 158, 127, 281, 276, 182, 147, 273, 148, 147, 152, 135, 134, 132, 307, 140, 255, 309, 128, 225, 295, 157, 142, 140, 151, 139, 249, 291, 305, 232, 132, 130, 291, 130, 246, 143, 129, 255, 155, 265, 279, 284, 160, 251, 133, 132, 158, 135, 216, 300, 138, 260, 132, 183, 250, 151, 292, 140, 221, 138, 132, 133, 283, 129, 249, 257, 245, 155, 256, 152, 267, 151, 239, 306, 263, 147, 278, 154, 143, 155, 160, 304, 240, 305, 136, 132, 301, 243, 147, 239, 260, 296, 129, 132, 135, 158, 259, 160, 278, 284, 274, 164, 148, 283, 304, 278, 287, 285, 226, 259, 135, 219, 256, 254, 153, 228, 228, 128, 153, 156, 294, 219, 153, 151, 155, 158, 245, 131, 225, 286, 135, 153, 252, 149, 258, 158, 149, 236, 239, 159, 255, 244, 134, 270, 255, 151, 135, 146, 254, 289, 156, 156, 128, 275, 155, 159, 148, 228, 298, 261, 254, 192,

In [ ]:
"""Проверим гипотезу, что сотрудники с высоким окладом
проводят на работе больше времени, чем сотрудники с низким
окладом"""
import scipy.stats as st

stat, p = st.ttest_ind(list_high, list_low)

print(f"Статистика = {stat:.5f}, p = {p:.10f}")

if p > 0.05:
    print('Не отклоняем нулевую гипотезу, средние, вероятно, одинаковые')
else:
    print('Отклоняем нулевую гипотезу, средние, вероятно, различаются')

Статистика = -0.72887, p = 0.4661023478
Не отклоняем нулевую гипотезу, средние, вероятно, одинаковые


Так как мы не отклюняем нулевую гипотезу, можно предположить, что сотрудники с высоким окладом не проводят больше времени на работе в статистически значимом смысле, чем сотрудники с низким окладом

6) Посмотрим распределение сотрудников по зарплатам в каждом департаменте по отдельности

In [ ]:
df_survey_abbrev = df_survey[['department', 'salary', 'satisfaction_level']]
df_survey_abbrev

department  salary  satisfaction_level
0          sales     low                0.38
1          sales  medium                0.80
2          sales  medium                0.11
3          sales     low                0.72
4          sales     low                0.37
...          ...     ...                 ...
14994    support     low                0.40
14995    support     low                0.37
14996    support     low                0.37
14997    support     low                0.11
14998    support     low                0.37

[14999 rows x 3 columns]

In [ ]:
df_survey_abbrev.columns = ['department', 'salary', 'workers']
df_survey_abbrev

department  salary  workers
0          sales     low     0.38
1          sales  medium     0.80
2          sales  medium     0.11
3          sales     low     0.72
4          sales     low     0.37
...          ...     ...      ...
14994    support     low     0.40
14995    support     low     0.37
14996    support     low     0.37
14997    support     low     0.11
14998    support     low     0.37

[14999 rows x 3 columns]

In [ ]:
'''посмотрим распределение сотрудников по зарплатам в каждом департаменте по отдельности'''
df_survey_sal = df_survey_abbrev.groupby(['department', 'salary']).agg({'workers': ['count']}).reset_index()
df_survey_sal

department  salary workers
                          count
0            IT    high      83
1            IT     low     609
2            IT  medium     535
3         RandD    high      51
4         RandD     low     364
5         RandD  medium     372
6    accounting    high      74
7    accounting     low     358
8    accounting  medium     335
9            hr    high      45
10           hr     low     335
11           hr  medium     359
12   management    high     225
13   management     low     180
14   management  medium     225
15    marketing    high      80
16    marketing     low     402
17    marketing  medium     376
18  product_mng    high      68
19  product_mng     low     451
20  product_mng  medium     383
21        sales    high     269
22        sales     low    2099
23        sales  medium    1772
24      support    high     141
25      support     low    1146
26      support  medium     942
27    technical    high     201
28    technical     low    1372
29    technical  medium    1147

Мы видим, что, в большинстве случаев, количество людей с высокой зарплатой в каждом отдельном департаменте - самое низкое (кроме департамента management). Чаще всего самое большое количество работников отдельного департамента получает низкую зарплату.

8) Рассчитаем следующие показатели среди уволившихся и не
уволившихся сотрудников (по отдельности):
● Доля сотрудников с повышением за последние 5 лет
● Средняя степень удовлетворенности
● Среднее количество проектов

In [ ]:
df_survey[['left', 'promotion_last_5years', 'satisfaction_level', 'number_project']].pivot_table(index='left')

number_project  promotion_last_5years  satisfaction_level
left                                                           
0           3.786664               0.026251            0.666810
1           3.855503               0.005321            0.440098

Мы видим, что среди уволившихся сотрудников степень удовлетворенности работой и доля сотрудников с повышением - ниже, чем среди не уволившихся. При этом проектов уволившиеся сделали чуть больше

9) Разделим данные на тестовую и обучающую выборки. Построим модель LDA, предсказывающую уволился ли сотрудник на основе имеющихся факторов (кроме department и salary). Также оценим качество модели на тестовой выборке

In [ ]:
df_survey_x = df_survey[['satisfaction_level', 'last_evaluation', 'number_project', 'average_montly_hours', 'time_spend_company', 'Work_accident', 'left', 'promotion_last_5years']]

In [ ]:
df_survey_x.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years  
0                   3              0     1                      0  
1                   6              0     1                      0  
2                   4              0     1                      0  
3                   5              0     1                      0  
4                   3              0     1                      0

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X = df_survey_x[['satisfaction_level', 'last_evaluation', 'number_project', 'average_montly_hours', 'time_spend_company', 'Work_accident', 'promotion_last_5years']]
y = df_survey_x['left']

In [ ]:
import numpy as np 
y = np.array(y)

In [ ]:
#приведем данные к единому масштабу
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.30, random_state=42)

In [ ]:
X_train.shape

(10499, 7)

In [ ]:
X_test.shape

(4500, 7)

In [ ]:
y_train.shape

(10499,)

In [ ]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

In [ ]:
lda = LinearDiscriminantAnalysis()
# обучаем данные
lda.fit(X_train, y_train)

LinearDiscriminantAnalysis()

In [ ]:
# делаем прогноз на тестовой выборке
lda.predict(X_test)

array([0, 0, 0, ..., 0, 0, 0])

In [ ]:
# смотрим разницу факта и прогноза
result = pd.DataFrame([y_test, lda.predict(X_test)]).T

In [ ]:
result.head(30)

0  1
0   0  0
1   0  0
2   0  0
3   1  0
4   0  0
5   0  0
6   0  0
7   0  1
8   0  0
9   0  1
10  1  0
11  0  0
12  0  0
13  0  0
14  1  0
15  0  0
16  0  0
17  0  0
18  0  0
19  0  0
20  0  0
21  1  0
22  1  1
23  0  0
24  1  0
25  1  0
26  1  0
27  0  0
28  0  0
29  0  1

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, lda.predict(X_test))
#проверяем точность прогноза

0.7557777777777778

С помощью модели мы предсказали увольнение сотрудников компании с точностью более 75%